### PR TITLE
Add driver support for ADC1 of stm32

### PR DESCRIPTION
**Pull request by cybertale**
This PR contains pinmux, dts and driver support for ADC1 for all series of stm32 that zephyr supports.

### CODE REVIEW

**Comment by cybertale** [comment]
To test this code, we should enable adc driver in menuconfig and add pinmux config in board specific file.
For f4 board, my configuration is 
```
#ifdef CONFIG_ADC0_CHAN0
	{STM32_PIN_PA0, STM32F4_PINMUX_FUNC_PA0_ADC123_IN0},
#endif
```

Test application code is
```/*
 * Copyright (c) 2012-2014 Wind River Systems, Inc.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <adc.h>
#include <zephyr.h>
#include <misc/printk.h>

uint16_t temp_data[8];

static const struct adc_channel_cfg channel_cfg_0 = {
        .channel_id       = 0,
};

const struct adc_sequence sequence = {
        .channels    = BIT(0),
        .buffer      = temp_data,
        .buffer_size = sizeof(temp_data),
        .resolution  = 12,
};

void main(void)
{
        int ret;
        struct device *adc_dev = device_get_binding(DT_ADC_0_NAME);
        adc_channel_setup(adc_dev, &channel_cfg_0);

        while (1) {
                ret = adc_read(adc_dev, &sequence);
                if (ret != 0) {
                        printk("adc_read() failed with code %d", ret);
                } else {
                        printk("adc_read() value %d\n", temp_data[0]);
                }
                k_sleep(1000);
        }
}
```

**Comment by cybertale** [comment]
Hi @erwango , I'v only test this code on f4 board, could you test it on other boards that available for you when you got some time?

**Comment by cybertale** [comment]
Generic DMA driver of stm32 is under development in #13364 , without DMA support many stm32s won't be able to catch up EOC interrupts generated by channels conversion, this will cause overrun error and lost data. So Only one channel can be put in one adc sequence in this driver.

**Comment by KwonTae-young** [comment]
I was interested in the ADC driver for STM32.
I tested it on [ST STM32F4DISCOVERY](https://docs.zephyrproject.org/latest/boards/arm/stm32f4_disco/doc/index.html) as shown below.
``` bash
west init zephyrproject -m https://github.com/KwonTae-young/zephyr/ --mr adc_stm32-new_stm32f4-disco
cd zephyrproject/zephyr
source zephyr-env.sh
mkdir samples/basic/adc/build
cmake -DBOARD=stm32f4_disco ..
make -j8 flash
```

ADC works well.
```
********** Booting Zephyr OS zephyr-v1.13.0-4846-g2eee90a41b *****
adc_read() value 1128
adc_read() value 952
adc_read() value 927
adc_read() value 1083
adc_read() value 916
adc_read() value 904
```

**Comment by cybertale** [comment]
@KwonTae-young Thanks for your interest. :grin:

File: drivers/adc/adc_stm32.c
Comment (anangl) [inline drivers/adc/adc_stm32.c]: You need to correct the locking scheme. Currently if an invalid sequence is requested (with an invalid resolution, for instance), the driver will remain locked after reporting the error.
See for instance how it is done in [adc_nrfx_adc](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_nrfx_adc.c#L214) (the [recent commit](https://github.com/zephyrproject-rtos/zephyr/commit/0906a51dacc947c5fa19a80e208549f9039556cd#diff-f8d824106434c62f9e6458df9242c4f5) that modified this driver contains a similar correction to what you need to apply) or in [adc_mcux_adc16](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_mcux_adc16.c#L150) that looks quite similar to your driver.

There is a test for ADC drivers in the Zephyr source tree that will help you check if the API implementation is correct. See [tests/drivers/adc/adc_api](https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/drivers/adc/adc_api).

File: drivers/adc/adc_stm32.c
Comment (anangl) [inline drivers/adc/adc_stm32.c]: The buffer pointer is not increased, so all the values read in a given sequence will be stored in the same place. Hence, the driver will work correctly only for simplest measurements (single channel, single sampling).
Again, try to run the [adc_api test](https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/drivers/adc/adc_api) to check if your driver behaves as expected.

File: drivers/adc/Kconfig.stm32
Comment (anangl) [inline drivers/adc/Kconfig.stm32]: `default n` is redundant.

**Review by anangl — CHANGES_REQUESTED** [review]

File: dts/arm/st/f0/stm32f0.dtsi
Comment (erwango) [inline dts/arm/st/f0/stm32f0.dtsi]: STM32_CLOCK_BUS_APB1_2

File: drivers/adc/Kconfig.stm32
Comment (erwango) [inline drivers/adc/Kconfig.stm32]: Rename to ADC1

File: drivers/adc/Kconfig.stm32
Comment (erwango) [inline drivers/adc/Kconfig.stm32]: Put whole section under "if ADC_1" and remove "depends on ADC_1"

File: drivers/adc/adc_stm32_v2.c
Comment (erwango) [inline drivers/adc/adc_stm32_v2.c]: Why not using `LL_ADC_Enable` and factorize this part?

**Review by erwango — CHANGES_REQUESTED** [review]
Initial feedback

File: drivers/adc/adc_stm32.c
Comment (cybertale) [reply drivers/adc/adc_stm32.c]: Hi @anangl , this is because currently we do not have a generic dma driver for stm32 and without dma, many stm32s won't be able to catch up every channel's EOC interrupt of a sequence, and overrun detection alerts. 
I'm working on dma driver on #13364, so before it's done, that's why only simple conversion is supported in this driver.
I'll try adc_api test, thank you.

File: drivers/adc/Kconfig.stm32
Comment (cybertale) [reply drivers/adc/Kconfig.stm32]: I really would like to do this while kconfig system of adc has a minor problem against this, which I described in #13501 . I will change them all to ADC_1 but then this should be merged after that PR.

File: drivers/adc/adc_stm32_v2.c
Comment (cybertale) [reply drivers/adc/adc_stm32_v2.c]: LL_ADC_Enable can be applied here, but I was thinking if I used LL lib somewhere, it's better to use LL lib everywhere. We talked about why I do not think LL lib is good for this generic driver, and except this function, some functions like getting EOC flags from registers uses different functions between different series of stm32 in LL lib.
So should I use LL and direct register access in the same driver?
I personally prefer to use direct register access for all operations.

File: drivers/adc/adc_stm32.c
Comment (cybertale) [reply drivers/adc/adc_stm32.c]: Sorry, I found this should be increased even in my situation during running the adc_api test, I'll change this.

**Comment by cybertale** [comment]
This PR should not be only merged until #13501 is merged.

File: drivers/adc/adc_stm32.c
Comment (anangl) [inline drivers/adc/adc_stm32.c]: Remove this line. The lock is now released in `adc_stm32_read[_sync]`.

Comment (cybertale) [reply drivers/adc/adc_stm32.c]: Right.

**Comment by cybertale** [comment]
Updated to use LL lib.

File: CODEOWNERS
Comment (erwango) [inline CODEOWNERS]: Please also add drivers/adc/adc_stm32.c

Comment (cybertale) [reply CODEOWNERS]: OK

**Comment by erwango** [comment]
@cybertale , since your work is close to mature and complete, could you update the description of the commit with a description of the work and the way it was tested? (this could edited later on with further addition)

**Comment by cybertale** [comment]
@erwango Sure!

File: tests/drivers/adc/adc_api/src/test_adc.c
Comment (anangl) [inline tests/drivers/adc/adc_api/src/test_adc.c]: These parameters are exactly the same as the ones for `NUCLEO_F091RC`. Wouldn't it be better to use:
```
#elif defined(CONFIG_BOARD_NUCLEO_F091RC) || \
      defined(CONFIG_BOARD_NUCLEO_F103RB)
```

Same applies to a couple of Nucleo boards using 10-bit resolution.

File: tests/drivers/adc/adc_api/src/test_adc.c
Comment (anangl) [inline tests/drivers/adc/adc_api/src/test_adc.c]: Duplicate entry.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

File: drivers/adc/Kconfig.stm32
Comment (erwango) [inline drivers/adc/Kconfig.stm32]: Miss ADC_2 and _3, since available in the driver

**Review by erwango — CHANGES_REQUESTED** [review]
This is promising, some more comments

File: tests/drivers/adc/adc_api/src/test_adc.c
Comment (cybertale) [reply tests/drivers/adc/adc_api/src/test_adc.c]: Yes, that's better

File: drivers/adc/Kconfig.stm32
Comment (cybertale) [reply drivers/adc/Kconfig.stm32]: I'm actually thinking about removing them, because there are so many different product lines, and different amount of ADCs they have, and also some ADCs share interrupt vector, so there should be some code for identifying which instance calls the interrupt. I would like another PR to do these things because this PR can be merged so users of STM32 can have basic ADC support.
But I'm worrying what Geo said about the merge window, but I see zephyr's doc says

> As a general rule, if you miss the merge window for a given feature, the best thing to do is to wait for the next development cycle. (An occasional exception is made for drivers for previously unsupported hardware; if they do not touch any other in-tree code, they cannot cause regressions and should be safe to add at any time).

So if this passes all tests and clean, we should be able to merge this PR soon, right?

**Comment by cybertale** [comment]
It seems like the Shippable tests failed because another work is under going, #14242 should be able to solve the problem.

File: boards/arm/nucleo_f401re/pinmux.c
Comment (anangl) [inline boards/arm/nucleo_f401re/pinmux.c]: Is this empty `#ifdef` block intentional?

**Review by anangl — APPROVED** [review]

File: boards/arm/nucleo_f401re/pinmux.c
Comment (cybertale) [reply boards/arm/nucleo_f401re/pinmux.c]: Sorry, my mistake

File: drivers/adc/Kconfig.stm32
Comment (erwango) [reply drivers/adc/Kconfig.stm32]: This is up to @galak to settle on this. Though, I'm not quite in favor of it. Even it could not be a source of regression, it could be a source of bugs, and branch maturity is tightly controlled.
Besides, there are several PR that could pretend be merged as this PR (L1/G0/WB). There is no reason this one is privileged. 
So I think the paragraph you mention should actually be corrected.

Comment (cybertale) [reply drivers/adc/Kconfig.stm32]: Understand, so I'll just wait. ;)

**Comment by GeorgeCGV** [comment]
@cybertale, tested on L452 SoC, works great.

File: drivers/adc/adc_stm32.c
Comment (shphilippe) [inline drivers/adc/adc_stm32.c]: k_busy_wait(LL_ADC_DELAY_INTERNAL_REGUL_STAB_US);

**Comment by shphilippe** [comment]
btw, I've added support for the STM32G0 in this branch: https://github.com/shphilippe/zephyr/tree/stm32g070-adc

File: drivers/adc/adc_stm32.c
Comment (cybertale) [reply drivers/adc/adc_stm32.c]: That's better.

**Comment by cybertale** [comment]
> btw, I've added support for the STM32G0 in this branch: https://github.com/shphilippe/zephyr/tree/stm32g070-adc

I see G0 support is not in plan of milestone v1.15, right? So you can feel free to push support for ADC of G0 after this. Or what's your plan? :)

**Comment by shphilippe** [comment]
I don't have much plans. Just trying to contribute back code which can be open sourced :)

File: drivers/adc/adc_stm32.c
Comment (erwango) [inline drivers/adc/adc_stm32.c]: Break a line here

File: drivers/adc/adc_stm32.c
Comment (erwango) [inline drivers/adc/adc_stm32.c]: Same here

File: drivers/adc/adc_stm32.c
Comment (erwango) [inline drivers/adc/adc_stm32.c]: Break a line here.

**Review by erwango — COMMENTED** [review]
Mostly nitpicks

**Comment by cybertale** [comment]
Hi @erwango , I added some comments and fixed style issues, how about this time?

**Review by erwango — APPROVED** [review]

**Comment by Odjek** [comment]
Hello. I tested the driver on stm32f413zh. It works well on PA0 but when I try to change the ADC pin it's not working anymore. Any ideas ? Thanks.

**Pull request merged**
